crates/bevy_tasks/src/task_pool.rs

```diff
@@ -347,13 +347,105 @@ impl TaskPool {
+                    let mut must_cancel_on_failure = Vec::with_capacity(spawned_ref.len());
+
+                    let mut completed_count = 0;
+                    let mut total_count = 0;
+                    loop {
+                        // collect any new tasks
+                        while let Ok(task) = spawned_ref.pop() {
+                            let index = tasks.len();
+                            tasks.push(task);
+                            results.push(None);
+
+                            // add to linked list of pending tasks
+                            nodes.push(Node {
+                                prev: None,
+                                next: head,
+                            });
+
+                            if let Some(head_index) = head {
+                                nodes[head_index].prev = Some(index);
+                            }
+                            head = Some(index);
+
+                            must_cancel_on_failure.push(true);
+                            total_count += 1;
+                        }
+
+                        // poll pending tasks for completion
+                        let mut next = head;
+                        while let Some(index) = next {
+                            if let Some(result) = future::poll_once(&mut tasks[index]).await {
+                                // result is ready
+                                if result.is_some() {
+                                    // task completed
+                                    results[index] = result;
+
+                                    // remove it from the pending list
+                                    let node = &mut nodes[index];
+                                    let prev = node.prev.take();
+                                    let next = node.next.take();
+
+                                    if let Some(p) = prev {
+                                        nodes[p].next = next;
+                                    } else {
+                                        // task was the head
+                                        head = next;
+                                    }
+
+                                    if let Some(n) = next {
+                                        nodes[n].prev = prev;
+                                    }
+
+                                    must_cancel_on_failure[index] = false;
+                                    completed_count += 1;
+                                } else {
+                                    // task failed
+                                    failure = true;
+                                    break;
+                                }
+                            }
+
+                            next = nodes[index].next;
+                        }
+
+                        if failure {
+                            // cancel any unfinished tasks that we've taken from the queue
+                            for (cancel, task) in
```

Comment:
I think you also need to check for tasks in spawned_ref

Reply:
I think the `Drop` impl still covers those.

Reply:
I think we should move the drop impl code into get results so all the canceling code is in one place